base_layer/core/src/proof_of_work/cuckaroo_pow.rs

```diff
@@ -0,0 +1,476 @@
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::collections::{HashMap, HashSet};
+
+use blake2::{
+    digest::{Update, VariableOutput},
+    Blake2bVar,
+};
+use thiserror::Error;
+
+use crate::{
+    blocks::BlockHeader,
+    proof_of_work::{siphash::siphash_block, Difficulty},
+};
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum CuckarooVerificationError {
+    #[error("PoW data contains non-zero padding")]
+    PowDataContainsNonZeroPadding,
+    #[error("PoW data is too short")]
+    PowDataTooShort,
+    #[error("Block header has an invalid PoW algorithm for Cuckaroo")]
+    BlockHeaderInvalidPowAlgorithm,
+    #[error("Nonce is too large")]
+    NonceTooLarge,
+    #[error("Nonces not ascending")]
+    NoncesNotAscending,
+    #[error("Endpoints don't match up")]
+    EndpointsDontMatch,
+    #[error("CycleTooShort")]
+    CycleTooShort,
+    #[error("CycleTooLong")]
+    CycleTooLong,
+    #[error("Edge already visited")]
+    EdgeAlreadyVisited,
+    #[error("Node has more than two edges")]
+    NodeHasMoreThanTwoEdges,
+    #[error("Cycle does not end at start")]
+    CycleDoesNotEndAtStart,
+    #[error("Cycle did not use all edges")]
+    CycleDidNotUseAllEdges,
+}
+
+pub fn cuckaroo_result(
+    header: &BlockHeader,
+    required_cycle_length: usize,
+    edge_bits: u8,
+) -> Result<Vec<u8>, anyhow::Error> {
+    let pow = header.pow.to_bytes();
+
+    // First byte must be 3 for Cuckaroo
+    if pow[0] != 3 {
+        return Err(anyhow::anyhow!(
+            CuckarooVerificationError::BlockHeaderInvalidPowAlgorithm
+        ));
+    }
+    let mut hasher = Blake2bVar::new(32).expect("Could not create Blake2bVar hasher");
+    hasher.update(&header.nonce.to_le_bytes());
+    hasher.update(header.mining_hash().as_slice());
+    let mut blob = Vec::with_capacity(32);
```

Comment:
This is safer as `fn finalize_variable(mut self, out: &mut [u8]) ` expects an initialised buffer.
```suggestion
    let mut blob = vec![0u8; 32];
```